geo/src/algorithm/concavehull.rs

```diff
@@ -0,0 +1,418 @@
+    /// and also uses ideas from the following paper:
+    /// www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf
+    ///
+    /// # Example
+    /// ```
+    /// use geo::{line_string, polygon};
+    /// use geo::algorithm::concavehull::ConcaveHull;
+    ///
+    /// // a square shape
+    /// let poly = polygon![
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    /// ];
+    ///
+    /// // The correct concave hull coordinates
+    /// let correct_hull = line_string![
+    ///     (x: 4.0, y: 0.0),
+    ///     (x: 4.0, y: 4.0),
+    ///     (x: 0.0, y: 4.0),
+    ///     (x: 0.0, y: 0.0),
+    ///     (x: 4.0, y: 0.0),
+    /// ];
+    ///
+    /// let res = poly.concave_hull(2.0);
+    /// assert_eq!(res.exterior(), &correct_hull);
+    /// ```
+    fn concave_hull(&self, concavity: T) -> Polygon<T>
+    where
+        T: Float + RTreeNum;
+}
+
+impl<T> ConcaveHull<T> for Polygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let points: Vec<_> = self.exterior().points_iter().collect();
+        Polygon::new(LineString::from(concave_hull(&points, concavity)), vec![])
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPolygon<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.exterior().0.iter().map(|c| Point(*c)))
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for LineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.clone().into_points(), concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiLineString<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        let aggregated: Vec<Point<T>> = self
+            .0
+            .iter()
+            .flat_map(|elem| elem.clone().into_points())
+            .collect();
+        Polygon::new(
+            LineString::from(concave_hull(&aggregated, concavity)),
+            vec![],
+        )
+    }
+}
+
+impl<T> ConcaveHull<T> for MultiPoint<T>
+where
+    T: Float + RTreeNum,
+{
+    fn concave_hull(&self, concavity: T) -> Polygon<T> {
+        Polygon::new(
+            LineString::from(concave_hull(&self.0.clone(), concavity)),
+            vec![],
+        )
+    }
+}
+
+fn find_point_closest_to_line<T>(
+    interior_points_tree: &RTree<Point<T>>,
+    line: Line<T>,
+    max_dist: &T,
+) -> Option<Point<T>>
+where
+    T: Float + RTreeNum,
+{
+    let h = max_dist.add(max_dist.clone());
```

Comment:
@michaelkirk Is there a way to avoid cloning here? When I take it out, I get the following error:
```
error[E0308]: mismatched types
   --> geo/src/algorithm/concavehull.rs:125:26
    |
117 | fn find_point_closest_to_line<T>(
    |                               - this type parameter
...
125 |     let h = max_dist.add(max_dist);
    |                          ^^^^^^^^
    |                          |
    |                          expected type parameter `T`, found `&T`
    |                          help: consider dereferencing the borrow: `*max_dist`
    |
    = note: expected type parameter `T`
                    found reference `&T`
    = help: type parameters must be constrained to match other types
    = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
```